src/main/scala/mesosphere/marathon/api/v2/AppTasksResource.scala

```diff
@@ -88,15 +81,46 @@ class AppTasksResource @Inject() (service: MarathonSchedulerService,
                 @PathParam("taskId") id: String,
                 @QueryParam("scale") scale: Boolean = false): Response = {
     val pathId = appId.toRootPath
+    def findToKill(appTasks: Set[MarathonTask]): Set[MarathonTask] = appTasks.find(_.getId == id).toSet
+
+    killTasks(pathId, findToKill, scale)
+  }
+
+  private def killTasks(pathId: PathId,
+                        findToKill: (Set[MarathonTask] => Set[MarathonTask]),
+                        scale: Boolean): Response = {
+    import mesosphere.util.ThreadPoolContext.context
+
     if (taskTracker.contains(pathId)) {
       val tasks = taskTracker.get(pathId)
-      tasks.find(_.getId == id).fold(unknownTask(id)) { task =>
-        service.killTasks(pathId, Seq(task), scale)
-        ok(Map("task" -> task))
+      val toKill = findToKill(tasks)
+      if (scale) {
+        val future = groupManager.group(pathId.parent).map {
+          case Some(group) =>
+            def updateAppFunc(current: AppDefinition) = current.copy(instances = current.instances - toKill.size)
+            val deploymentPlan = result(groupManager.updateApp(
+              pathId, updateAppFunc, Timestamp.now(), force = false, toKill = toKill))
+            deploymentResult(deploymentPlan)
+
+          case None => unknownGroup(pathId.parent)
+        }
+        result(future)
+      }
+      else {
+        // setting scale = false in order to make the differing behaviour explicit
+        service.killTasks(pathId, toKill, scale = false)
+      }
+
+      if (toKill.size == 1) {
+        ok(Map("task" -> toKill.head))
```

Comment:
I fear that your generalization does not work here because it might result in `deleteMany` returning a result with a `task` field if only on task was found. Also `deleteOne` might return an empty `tasks` list instead of `"task": null`.

Unfortunately, I cannot think of an elegant solution to this right now.